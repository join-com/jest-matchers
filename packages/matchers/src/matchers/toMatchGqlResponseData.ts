import { diff, matcherHint, printExpected, printReceived } from 'jest-matcher-utils'

export interface IGqlResponse {
  body: {
    singleResult: {
      data?: unknown
    }
  }
}

export const toMatchGqlResponseData = <E extends object>(
  response: IGqlResponse,
  expectedData: E,
): jest.CustomMatcherResult => {
  const responseData = response.body.singleResult.data
  const pass = responseMatches(responseData, expectedData)
  return {
    pass,
    message: () => {
      const hintString = matcherHint('.toMatchGqlResponseData')
      const expectedString = printExpected(expectedData)
      const mismatchString = formatMismatched(responseData, expectedData)
      return `${hintString}\n\n` + `Expected response data to match:\n${expectedString}\n` + mismatchString
    },
  }
}

const formatMismatched = (responseData: unknown, expectedData: unknown): string => {
  const diffString = diff(responseData, expectedData) ?? ''

  const receivedString = printReceived(responseData)
  const message = `${receivedString}\n` + 'Difference:\n' + `${diffString}`

  return message
}

const responseMatches = <E extends object>(responseData: unknown, expectedData: E): boolean => {
  try {
    expect(responseData).toMatchObject(expectedData)
    return true
  } catch (_e) {
    return false
  }
}
