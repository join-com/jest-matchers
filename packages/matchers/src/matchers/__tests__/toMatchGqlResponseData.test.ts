import { IGqlResponse, toMatchGqlResponseData } from '../toMatchGqlResponseData'

expect.extend({ toMatchGqlResponseData })

describe('toMatchGqlResponseData', () => {
  it('passes when response data is equal to expected', () => {
    const response: IGqlResponse = {
      body: {
        singleResult: {
          data: {
            a: 1,
            b: 2,
          },
        },
      },
    }

    expect(response).toMatchGqlResponseData({
      a: 1,
      b: 2,
    })
  })

  it('passes with response data matches expected', () => {
    const response: IGqlResponse = {
      body: {
        singleResult: {
          data: {
            a: 1,
            b: 2,
          },
        },
      },
    }

    expect(response).toMatchGqlResponseData({ a: 1 })
    expect(response).toMatchGqlResponseData({ b: 2 })
  })

  it('fails when response data do not match expected', () => {
    const response: IGqlResponse = {
      body: {
        singleResult: {
          data: {
            a: 1,
            b: 2,
          },
        },
      },
    }

    const validate = () => {
      expect(response).toMatchGqlResponseData({
        a: 3,
        b: 4,
      })
    }

    expect(validate).toThrowErrorMatchingSnapshot()
  })
})
