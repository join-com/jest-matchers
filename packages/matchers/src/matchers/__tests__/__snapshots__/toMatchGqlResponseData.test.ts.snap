// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toMatchGqlResponseData fails when response data do not match expected 1`] = `
"expect(received).toMatchGqlResponseData(expected)

Expected response data to match:
{"a": 3, "b": 4}
{"a": 1, "b": 2}
Difference:
- Expected
+ Received

  Object {
-   "a": 1,
-   "b": 2,
+   "a": 3,
+   "b": 4,
  }"
`;
